qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+""""Central management of pulse program parameters.
+
+Background
+==========
+
+In contrast to ``QuantumCircuit``, in pulse programs parameter objects can be stored in
+every places at different layers, for example
+
+- program's variable: ``ScheduleBlock.alignment_context._context_params``
+
+- instruction's operands: ``ShiftPhase.phase``, ...
+
+- operand's parameters: ``pulse.parameters``, ``channel.index`` ...
+
+This complexity is due to tight coupling of program to underlying device Hamiltonian,
+i.e. variance of physical parameters of qubits and their coupling.
+If we want to define a program that can be used with arbitrary qubits,
+we should be able to parametrize every control parameters in the program
+since they are basically qubit dependent, also varying from time to time.
+
+
+Implementation
+==============
+
+Managing parameters in each object consisting a program, i.e. ``ParameterTable`` model,
+makes the framework quite complicated. In the ``ParameterManager`` class this module provides,
+the parameter assignment operation is performed by the visitor instance.
+
+The visitor pattern is a way of separating data processing from an object on which it operates.
+This removes the overhead of parameter management from each piece of the program.
+Computational complexity of the parameter assignment operation may be increased
+from the parameter table model of ~O(1), however usually this calculation occurs
+only once before the program is executed. Thus this doesn't hurt user experience during
+pulse programming. On the contrary, it removes parameter table object and associated logic
+from each object, yielding smaller object creation cost and higher performance
+as the data amount scales.
+
+Note that we don't need to write any parameter management logic for each object,
+and thus this parameter framework gives greater scalability to the pulse module.
+"""
+from copy import deepcopy, copy
+from typing import List, Dict, Set, Any, Union
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parameterexpression import ParameterExpression, ParameterValueType
+from qiskit.pulse import instructions, channels
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.library import ParametricPulse, Waveform
+from qiskit.pulse.schedule import Schedule, ScheduleBlock
+from qiskit.pulse.transforms.alignments import AlignmentKind
+from qiskit.pulse.utils import format_parameter_value
+
+
+class NodeVisitor:
```

Comment:
This will be great! Consolidate all the craziness that was going on everywhere else:)